Source/WebCore/rendering/TextBoxPainter.cpp

```diff
@@ -1060,11 +1060,49 @@ void TextBoxPainter::paintPlatformDocumentMarkers()
 
 #if ENABLE(WRITING_TOOLS)
 
-#if USE(APPLE_INTERNAL_SDK)
-#import <WebKitAdditions/TextBoxPainterAdditions.cpp>
-#else
-static void drawUnifiedTextReplacementUnderline(GraphicsContext&, const FloatRect&, IntSize) { }
-#endif
+constexpr Seconds writingToolsAnimationLoop = 10000_ms;
+static void drawWritingToolsUnderline(GraphicsContext& context, const FloatRect& rect, IntSize frameSize)
+{
+    auto radius = rect.height() / 2.0;
+    auto minX = rect.x();
+    auto maxX = rect.maxX();
+    auto minY = rect.y();
+    auto maxY = rect.maxY();
+    auto midY = (minY + maxY) / 2.0;
+
+    auto frameX = frameSize.width();
+    auto frameY = frameSize.height();
+
+    constexpr auto redColor = SRGBA<uint8_t> { 227, 100, 136 };
+    constexpr auto yellowColor = SRGBA<uint8_t> { 242, 225, 162 };
+    constexpr auto purpleColor = SRGBA<uint8_t> { 154, 109, 209 };
+
+    auto animationProgress = (MonotonicTime::now() % writingToolsAnimationLoop).value() / 10;
+
+    auto xOffset = frameX * fmod(animationProgress + midY / frameY, 1.0);
+    constexpr std::array colorList = { purpleColor, redColor, yellowColor, redColor, purpleColor, purpleColor, redColor, yellowColor, redColor, purpleColor };
+
+    Ref gradient = Gradient::create(Gradient::LinearData { FloatPoint(0 - xOffset, 0), FloatPoint(frameX * 2 - xOffset, frameY) }, { ColorInterpolationMethod::SRGB { }, AlphaPremultiplication::Unpremultiplied });
+
+    auto colorStop = 0.f;
+    auto colorIncrement = 1.0 / colorList.size();
+    for (auto color : colorList) {
+        gradient->addColorStop({ 0.f + colorStop, color });
```

Comment:
Hmm why do we need to say `0.f + colorStop`, rather than just `colorStop`?

Reply:
I think it was left over from it not being a float or something.